[TypeInfo] Simple array should be array type

matijn-made opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | 
| License       | MIT

While using latest Symfony Live Components we detected that simple array docblocks (`bool[]`) were parsed as list type instead of array type. I couldn't find any documentation on what's officially known as list vs. array, as PHP does not really distinguish between list and array. It is my understanding hower that using `bool[]` means array. PHPstan docs also mentions `Type[]` being array (https://phpstan.org/writing-php-code/phpdoc-types)
I've added a few tests to verify the new behaviour.

I also added an extra check for which key types are supported for arrays (with tests). I'm not sure if these should go in this PR, or if I should add a second PR for this.

matijn-made committed:
[TypeInfo] Simple array should be array type

stof commented:
this change is legitimate. None of the static analysis tools treat `bool[]` as meaning a list.

mtarld commented on src/Symfony/Component/TypeInfo/Tests/TypeResolver/StringTypeResolverTest.php:
```diff
@@ -73,7 +72,17 @@ public static function resolveDataProvider(): iterable
+
+        // list
+        yield [Type::list(Type::bool()), 'list<bool>'];
+        yield [Type::list(Type::bool()), 'list<int, bool>'];
```
The `list` generic type only accepts one value. This should not work.

mtarld commented on src/Symfony/Component/TypeInfo/Tests/TypeResolver/StringTypeResolverTest.php:
```diff
@@ -253,4 +261,22 @@ public function testCannotResolveValueOfInvalidType()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->resolver->resolve('list<mixed, string>');
+    }
```
This must be updated to fail if a key has been defined IMHO.

mtarld commented on src/Symfony/Component/TypeInfo/Type/CollectionType.php:
```diff
@@ -49,6 +49,20 @@ public function __construct(
+                $types = $keyType->getTypes();
+            } else {
+                $types = [$keyType];
+            }
```
```suggestion
            $keyTypes = $keyType instanceof UnionType ? $keyType->getTypes() : [$keyType];
```

mtarld commented on src/Symfony/Component/TypeInfo/Type/CollectionType.php:
```diff
@@ -49,6 +49,20 @@ public function __construct(
+            }
+
+            foreach ($types as $type) {
+                if ($type instanceof BuiltinType && !\in_array($type->getTypeIdentifier(), [TypeIdentifier::INT, TypeIdentifier::STRING], true)) {
```
What if `$type` is not a `BuiltinType`?

mtarld replied:
```suggestion
                if (!$type instanceof BuiltinType || !\in_array($type->getTypeIdentifier(), [TypeIdentifier::INT, TypeIdentifier::STRING], true)) {
```

matijn-made commented:
Thank you all for your comments. I have addressed them in the latest commit. 
The tests are also updated and green. 
The low//high deps tests are failing because they install different version of TypeInfo instead of using the changes in this PR. Is it oke for these unit tests to fail now?

nicolas-grekas commented:
The low-deps should be fixed by bumping the minimum version in framework-bundle

nicolas-grekas commented on src/Symfony/Component/PropertyInfo/composer.json:
```diff
@@ -26,7 +26,7 @@
         "symfony/deprecation-contracts": "^2.5|^3",
         "symfony/string": "^6.4|^7.0",
-        "symfony/type-info": "^7.3.5"
+        "symfony/type-info": "^7.3.8|^7.4"
```
please rebase your branch after fetching 7.3 to get rid of the conflict

xabbuh commented on src/Symfony/Component/JsonStreamer/Tests/Fixtures/Model/DummyWithArray.php:
```diff
@@ -4,7 +4,7 @@
 class DummyWithArray
 {
-    /** @var ClassicDummy[] */
+    /** @var list<ClassicDummy> */
```
I am not sure if it is actually a good idea to do this change or if we shouldn’t introduce additional `DummyWithList` and `DummyWithNestedList` fixtures instead to cover both cases.

matijn-made replied:
I also doubted about that change. I have reverted it and added a second test for list type

mtarld commented on src/Symfony/Component/TypeInfo/Type/CollectionType.php:
```diff
@@ -25,7 +25,7 @@
  * @implements WrappingTypeInterface<T>
  */
-class CollectionType extends Type implements WrappingTypeInterface
+class CollectionType extends Type implements \Stringable, WrappingTypeInterface
```
```suggestion
class CollectionType extends Type implements WrappingTypeInterface
```
`Type` is already `Stringable`.

matijn-made replied:
I blindly copied Fabbot suggestion, but you're absolutely right. Can I just ignore fabbot?

mtarld replied:
In that case, yes I think so

mtarld commented on src/Symfony/Component/JsonStreamer/Tests/JsonStreamWriterTest.php:
```diff
@@ -121,21 +122,21 @@ public function testWriteCollection()
         );
 
-        $dummyWithNestedArray1 = new DummyWithNestedArray();
+        $dummyWithNestedArray1 = new DummyWithNestedList();
```
Can vars be renamed as well? 🙂

matijn-made replied:
I think it makes more sense to test both nested array and nested list scenario, so I've added those tests as well. (with correct variable names)
